code/game/objects/structures/barsign.dm

```diff
@@ -41,120 +53,463 @@
+	if(build_stage == BARSIGN_WIRED && !istype(current_sign, /datum/barsign/hiddensigns/building/wired))
+		set_sign(new /datum/barsign/hiddensigns/building/wired)
+	if(build_stage == BARSIGN_COMPLETE)
+		if((stat & BROKEN) && !istype(current_sign, /datum/barsign/hiddensigns/signbroken))
+			set_sign(new /datum/barsign/hiddensigns/signbroken)
+			return
+		if((stat & NOPOWER) && !(stat & BROKEN) && !istype(current_sign, /datum/barsign/hiddensigns/signoff))
+			set_sign(new /datum/barsign/hiddensigns/signoff)
+			return
+		if((stat & EMPED) && !(stat & BROKEN|NOPOWER) && !istype(current_sign, /datum/barsign/hiddensigns/empbarsign))
+			set_sign(new /datum/barsign/hiddensigns/empbarsign)
+			return
+	if(!current_sign)
+		turn_off()
 		return
 	icon_state = current_sign.icon
 
-/obj/structure/sign/barsign/update_overlays()
+// Allows the sign to be visible in complete darkness.
+/obj/machinery/barsign/update_overlays()
 	. = ..()
 	underlays.Cut()
 
-	if(current_sign == "empty" || broken || !current_sign)
+	if(power_state < ACTIVE_POWER_USE || stat & (BROKEN|NOPOWER) || !current_sign || build_stage < BARSIGN_COMPLETE)
 		return
 
 	underlays |= emissive_appearance(icon, current_sign.icon)
 
-/obj/structure/sign/barsign/obj_break(damage_flag)
-	if(!broken && !(flags & NODECONSTRUCT))
-		broken = TRUE
-		prev_sign = current_sign
-		set_light(0)
-		update_icon()
-
-/obj/structure/sign/barsign/deconstruct(disassembled = TRUE)
-	new /obj/item/stack/sheet/metal(drop_location(), 2)
-	new /obj/item/stack/cable_coil(drop_location(), 2)
-	qdel(src)
-
-/obj/structure/sign/barsign/play_attack_sound(damage_amount, damage_type = BRUTE, damage_flag = 0)
-	switch(damage_type)
-		if(BRUTE)
-			playsound(src.loc, 'sound/effects/glasshit.ogg', 75, TRUE)
-		if(BURN)
-			playsound(src.loc, 'sound/items/welder.ogg', 100, TRUE)
+/obj/machinery/barsign/examine(mob/user)
+	. = ..()
+	switch(build_stage)
+		if(BARSIGN_FRAME)
+			. += "<span class='notice'>It's missing a <i>circuit board</i> and the <b>bolts</b> are exposed.</span>"
+		if(BARSIGN_CIRCUIT)
+			. += "<span class='notice'>The frame needs <i>wiring</i> and the circuit board could be <b>pried out</b>.</span>"
+		if(BARSIGN_WIRED)
+			. += "<span class='notice'>The frame lacks a <i>glass screen</i> and is filled with wires that could be <b>cut</b>.</span>"
+		if(BARSIGN_COMPLETE)
+			. += "<span class='info'><b>Alt-Click</b> to toggle it's power.</span>"
+			if(panel_open)
+				. += "<span class='notice'>It is disabled by it's <i>unscrewed</i> maintenance panel that exposes an area from which the screen could be <b>pried out</b>.</span>"
+
+/obj/machinery/barsign/proc/is_on()
+	if(power_state == ACTIVE_POWER_USE)
+		return TRUE
+	return FALSE
+
+/obj/machinery/barsign/proc/toggle_on_off()
+	if(is_on())
+		return turn_off()
+	else
+		return turn_on()
+
+/obj/machinery/barsign/proc/turn_off()
+	if((stat & (BROKEN|EMPED|MAINT)) || !is_on() || build_stage < BARSIGN_COMPLETE)
+		return FALSE
+	save_sign()
```

Comment:
Why not just give the current_sign argument here instead of doing that "datum/barsign/S=current_sign" over the save_sign?